application/client/src/app/ui/tabs/sources/multiplefiles/component.ts

```diff
@@ -0,0 +1,201 @@
+import { detectFileType } from './file.holder';
+import { TabControls } from '@service/session';
+import { bytesToStr } from '@env/str';
+import { EContextActionType, IContextAction } from './structure/component';
+
+@Component({
+    selector: 'app-tabs-source-multiple-files',
+    templateUrl: './template.html',
+    styleUrls: ['./styles.less'],
+})
+@Initial()
+@Ilc()
+export class TabSourceMultipleFiles extends Holder implements AfterContentInit, OnDestroy {
+    @Input() files!: File[];
+    @Input() tab!: TabControls;
+
+    @HostListener('window:keydown', ['$event'])
+    handleKeyDown(event: KeyboardEvent) {
+        if (this.filter.keyboard(event)) {
+            this.matcher.search(this.filter.value());
+            this.filesUpdate.emit(
+                this.state.files
+                    .sort((a: FileHolder, b: FileHolder) => b.getScore() - a.getScore())
+                    .filter((file: FileHolder) => file.getScore() > 0),
+            );
+        }
+    }
+
+    public filter: Filter = new Filter(this.ilc());
+    public state: State = new State();
+    public filesUpdate: Subject<FileHolder[]> = new Subject();
+
+    private _selectedTypes: FileType[] = [];
+
+    constructor() {
+        super();
+    }
+
+    public ngAfterContentInit() {
+        const state = this.tab.storage<State>().get();
+        if (state !== undefined) {
+            this.state = state;
+        } else {
+            this.state.files = this.files.map(
+                (file: File) =>
+                    new FileHolder(this.matcher, {
+                        filename: file.filename,
+                        modificationDate: file.stat.mtimeMs,
+                        name: file.name,
+                        path: file.path,
+                        size: file.stat.size,
+                        type: file.type,
+                    }),
+            );
+            this.tab.storage().set(this.state);
+        }
+        this._countAndCheck();
+    }
+
+    public ngCancel() {
+        this.tab.close();
+    }
+
+    public ngOnDestroy() {
+        this.tab.storage<State>().set(this.state);
+    }
+
+    public ngConcat() {
+        // TODO - Concatenate files
+        // TODO - Close tab - if DLT new Tab
+    }
+
+    public ngOnDestination() {
+        // TODO - Open file explorer to select path
+    }
+
+    public ngOnDrop(fileList: any) {
+        for (let i = 0; i < fileList.length; i++) {
+            const item = fileList.item(i);
+            if (item !== null) {
+                this.ilc()
+                    .services.system.bridge.files()
+                    .getByPath([item.path])
+                    .then((result) => {
+                        if (result.length !== 1) {
+                            this.log().error(`Invalid file stat info`);
+                            return;
+                        }
+                        if (item != null) {
+                            this.state.files.push(
```

Comment:
you check two time `item !== null`. To make Typescript happy and have single check you can reassign the value:
```
if (item === null) {
    continue;
}
let file = item;
// do all stuff
```
As soon as you created new variable `file` you don't need to double check in in callbacks bellow in promisses.

Reply:
In general I would reorganize a little this method.
1. Here looks pretty suitable usage of `Promise.all` to understand when all files processed
2. I would add progress bar to indicate start and finish processing

Reply:
BWT `getByPath(filenames: string[])` accept an array of files, well you can get info for all just with one call, so no need `Promise.all`, no need loop